### PR TITLE
Enable logging for live tests

### LINKED ISSUE #11981: Add shared file logger implementation to the test framework
We need a common logger implementation that would always log requests/responses into files that would be added as an artifact.

There should be one file per test invocation with the path something like:

`Azure.Core/HttpClientTransportTests/SomeTestNameAsync.txt`

This would allow us to get rid of multiple test logger implementations and always have some logs when a test unexpectedly fails without requiring someone to instrument the code.

cc @heaths @JoshLove-msft

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Moves the TestLogger from Search to TestFramework, and logs for all tests instead of just failed.
Also deletes the logger implementation from Storage.
For passed tests, the attachments won't be able to be viewed directly in Dev Ops like they are for failed tests. Instead, you download the test output from the parent test run, and you can view the output for each test in VS.

Partially fixes https://github.com/Azure/azure-sdk-for-net/issues/11981 - doesn't cover hanging or crashed test runs.

### CODE REVIEW

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 1]
```diff
@@ -71,6 +103,7 @@ public virtual void StopTestRecording()
             save |= SaveDebugRecordingsOnFailure;
 #endif
             Recording?.Dispose(save);
+            Logger?.OutputEventsForTest();
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:106]: Why don't we directly log to TestContext.Log?

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs [hunk 2]
```diff
@@ -50,6 +51,36 @@ private string GetSessionFilePath(string name = null)
...
+        private static TestLogger Logger { get; set; }
+
+        /// <summary>
+        /// Start logging events to the console if debugging or in Live mode.
+        /// This will run once before any tests.
+        /// </summary>
+        [OneTimeSetUp]
+        public void StartLoggingEvents()
+        {
+            if (Debugger.IsAttached || Mode == RecordedTestMode.Live)
```
Comment (tg-msft) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:67]: Nit - flip the order of the two conditions.  `Debugger.IsAttached` is probably a lot more expensive to check.

File: sdk/core/Azure.Core.TestFramework/src/TestLogger.cs [hunk 3]
```diff
@@ -56,15 +51,16 @@ internal class TestLogger : IDisposable
...
+        /// can be opened in VS to review the logs.
         /// This will include the HTTP requests and responses.
         /// </summary>
         public void OutputEventsForTest()
         {
-            if (_verbose || TestContext.CurrentContext.Result.Outcome.Status == TestStatus.Failed)
-            {
-                TestContext.Out.WriteLine(_log.ToString());
-            }
+            TestContext.Out.WriteLine(_log.ToString());
```
Comment (tg-msft) [inline sdk/core/Azure.Core.TestFramework/src/TestLogger.cs:63]: Why the change to log everything?  Did the DevOps experience here get a radical improvement?

**Review by tg-msft — APPROVED** [review]

File: sdk/core/Azure.Core.TestFramework/src/TestLogger.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/TestLogger.cs:63]: I'm not sure if DevOps changed, but the passed tests will not be included in the console or in individual test attachments. You can download all the test output from the parent test run though.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:106]: Do you mean do away with the Logger abstraction?

File: sdk/core/Azure.Core.TestFramework/src/TestLogger.cs (hunk 3 above)
Comment (heaths) [reply sdk/core/Azure.Core.TestFramework/src/TestLogger.cs:63]: Because for potential shared resource issues, we need successful logs to see who created / updated / deleted it that cause other tests to fail.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs (hunk 1 above)
Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:106]: No, keep the abstraction just lose the StringBuilder

File: sdk/core/Azure.Core.TestFramework/src/TestLogger.cs (hunk 3 above)
Comment (heaths) [inline sdk/core/Azure.Core.TestFramework/src/TestLogger.cs:63]: Is this any better than just logging to stdout? I know we talked about problems with that, but I thought `TestContext.Out` had the same problems. If not better, it's pointless to maintain a separate logger (than the console logger) that uses increased memory and time to reformat the payload and store it in an expanding `StringBuilder`, just to attach it as stdout anyway.

**Review by heaths — APPROVED** [review]

**Comment by heaths** [comment]
While not directly tied to this change, it amplifies the problem of having all the TRX files have the same name. This can be customized in the eng/pipelines/templates/jobs/archetype-sdk-tests.yml to, perhaps, add the platform to the `-logger trx:LogFileName=$(TestTargetFramework).trx` line so we can easily extract, compare, and identify the different TRX files.

File: sdk/core/Azure.Core.TestFramework/src/TestLogger.cs (hunk 3 above)
Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/TestLogger.cs:63]: Logging to STDOUT is going through another layer of indirection just to get to `TestContext.Out`. 

We should keep the logger type because we are going to add more stuff to it (like outputting to a file)

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs (hunk 1 above)
Comment (heaths) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:106]: We'll need the `StringBuilder` to attached failed test output to the test if we're not logging to stdout. While having to look over successful tests' output is hopefully rare, looking at failed tests' output in the UI is going to be much more common. Please keep this the way it is. We can tee to a file for all failed and successful tests and let it write to the TRX file.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs]: I'm not sure I follow. We will always be writing to TestContext.Out, regardless of whether test passed or failed. For passed tests, this will not be attached to the individual tests within dev ops, but you can download the directory of trx files to see the output.

Comment (heaths) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs]: Seems we were referring to different StringBuilders.

File: sdk/core/Azure.Core.TestFramework/src/TestLogger.cs [hunk 4]
```diff
@@ -27,44 +27,14 @@ internal class TestLogger : IDisposable
...
-        /// OutputEventsForTest.
-        /// </summary>
-        private StringBuilder _log;
-
-        /// <summary>
-        /// Gets a value indicating whether or not to log output for successful tests.
-        /// This could be treated as a numeric level in the future if we wanted different
-        /// levels of verbosity, but for now is only boolean.
-        /// </summary>
-        private readonly bool _verbose = Environment.GetEnvironmentVariable("AZURE_SEARCH_TEST_VERBOSE") != null;
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/TestLogger.cs:42]: I think @heaths added this variable to some pipeline somewhere, we need to remove it.

Comment (heaths) [reply sdk/core/Azure.Core.TestFramework/src/TestLogger.cs:42]: It's in sdk/seach/tests.yml in the `EnvVars` template parameter.

**Comment by JoshLove-msft** [comment]
> While not directly tied to this change, it amplifies the problem of having all the TRX files have the same name. This can be customized in the eng/pipelines/templates/jobs/archetype-sdk-tests.yml to, perhaps, add the platform to the `-logger trx:LogFileName=$(TestTargetFramework).trx` line so we can easily extract, compare, and identify the different TRX files.
> 
> /cc @weshaggard for comment on changing this. I doubt it would have any negative impact.

Updated

**Comment by pakrym** [comment]
What's up with cancelled test runs?

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml [hunk 5]
```diff
@@ -96,7 +96,7 @@ jobs:
         condition: always()
         displayName: "Publish Results ($(TestTargetFramework))"
         inputs:
-          testResultsFiles: "**/$(TestTargetFramework)*.trx"
+          testResultsFiles: "**/$(OSName)-$(TestTargetFramework)*.trx"
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Are there multiple trx files any longer after this change? In a similar question do we still need the mergeTestResults: true option below? Perhaps all you are really after is to set that parameter to false?

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Yes there are separate trx files for each platform. Actually there is a trx file per project for each platform. Each trx file now has the OSName prepended. 
My understanding is that mergeTestResults makes it so that each of the trx files (1 for each test project) will be included under the same run (1 for each platform). If this was false, I think we would have multiple parent test runs per platform (1 for each test project).

Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Nevermind there will still be one per test library. 

However I'm still a little unclear why it is interesting to change this name. Do we plan to download the trx files directly?

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: The use case is for inspecting the output of passed tests. The output doesn't get included in the Test tabs for passed tests.

**Comment by JoshLove-msft** [comment]
> What's up with cancelled test runs?

https://dev.azure.com/azure-sdk/public/_build/results?buildId=389640&view=logs&j=9cfc3b0b-53b1-524f-85ed-0ef4880bef82

Looks like a transient failure.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml (hunk 5 above)
Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Yeah library name would be nice to add as well.

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Is it possible to add this? I thought this run is for the whole service directory so we wouldn't have a chance to alter the name based on test project. Is there a predefined variable for this?

Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Good point. Without running the dotnet test command on the individual projects we cannot easily give it a different file name per library. However I will say when I downloaded the trx zip file from the published test results they all have the same name and so you have to carefully copy each out if you want to look at them individually.

Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Based on https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-vstest 

>If LogFileName isn't provided, a unique file name is created to hold the test results.

So I wonder if we should just not specify the trx file name. Sure we might get a random file name but hopefully it will be unique for each library.

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: But then we wouldn't get the OS name. That said, I'm not sure of the value of the OS name given that I think you have to download the test runs one at a time. You can name the directory anything you want after downloading.
@heaths thoughts?

Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: You will have to click on the "OS - TestFramework" link in the uploaded test log page and then download the trx.zip file so you will know based on that. I'd suggest giving a test in your PR (or another) and see what the end result looks like.

Comment (heaths) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: I requested this and would like this format because often times - to diagnose potential shared resources - I need to examine multiple logs at once. If they have the same names, it's a hassle during extraction. There are workarounds, of course, but just extracting them all to the same directory is easiest. Is there some reason we shouldn't take this change?

Comment (heaths) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Also, if you want to keep the filter simple, you could just change the file name to "*.trx" under the same directory. It's not like there'll be any others (machines are hosted and reclaimed after each run).

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: The issue Wes mentioned is that the trx files will be the same for each test project. I'm not sure this is much of an issue outside of Storage.

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: e.g. here is what it looks like when downloading storage test run for one platform.

image

Comment (heaths) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:99]: Yeah, KV has that problem too but, as you two mentioned above (I tried to solve this and couldn't find a good solution either), there's not much we can do about that. Maybe letting the file name be random would be better. The file name isn't immediately identifiable, but it'll be in the root element (or somewhere under that toward the top - don't remember exactly off hand).

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml]: Here is what it looks like with no filename specified. Note the prefix differs between OS, but the suffix is just the time stamp. It looks like the only thing we lose as compared to using OSName is differentiated between Core and Framework on Windows.

Linux:
image

Windows Core:
image

Windows Framework:
image

Mac:
image

Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml]: At least that would allow us to always extract the zip without conflicts and file overwrites. We will just need to ensure the zip we download is extracted into a specifically named directory based on the OS and framework.

Comment (JoshLove-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml]: @pakrym @heaths are we good to merge?

**Review by weshaggard — APPROVED** [review]
